source/common/tcp_proxy/upstream.h

```diff
@@ -89,10 +89,27 @@ class TcpUpstream : public GenericUpstream {
   Tcp::ConnectionPool::ConnectionDataPtr upstream_conn_data_;
 };
 
+struct GenericPoolReadyDeferrer {
+  std::unique_ptr<GenericUpstream> upstream_;
+  GenericConnectionPoolCallbacks* callbacks_{};
+  Upstream::HostDescriptionConstSharedPtr host_;
+  Ssl::ConnectionInfoConstSharedPtr ssl_info_;
+
+  void call(Http::RequestEncoder& request_encoder) {
+    ASSERT(callbacks_ != nullptr);
+    callbacks_->onGenericPoolReady(nullptr, std::move(upstream_), host_,
+                                   request_encoder.getStream().connectionLocalAddress(), ssl_info_);
+  }
+  void release() {
+    if (upstream_) {
```

Comment:
should we also call onGenericPoolFailure here?
Basically if we defer the ready call, and something unanticipated happens, we want to make the session isn't left hanging.